src/plugins/lightscript.js

```diff
@@ -65,64 +65,62 @@ pp.rewriteOperator = function (node) {
   }
 };
 
-pp.parseForFrom = function (node, iterator) {
-  this.expectContextual("from");
-  // `for i from`
-  const arrayOrRangeStart = this.parseExpression(true);
-  if (this.match(tt._thru) || this.match(tt._til)) {
-    return this.parseForFromRange(node, iterator, arrayOrRangeStart);
+function getIterableForKeyword(name) {
+  if (name === "idx" || name === "elem") {
+    return "array";
+  } else if (name === "key" || name === "val") {
+    return "object";
   } else {
-    return this.parseForFromArray(node, iterator, null, arrayOrRangeStart);
+    return null;
   }
-};
+}
 
-pp.parseForFromArray = function (node, iterator, element = null, array = null) {
-  if (iterator.type !== "Identifier") this.unexpected(iterator.start);
-  if (element && element.type !== "Identifier") this.unexpected(element.start);
-  node.id = iterator;
-  node.elem = element;
+pp.parseEnhancedForIn = function (node) {
+  const firstIterationType = this.state.value;
+  const iterableType = getIterableForKeyword(firstIterationType);
 
-  if (array) {
-    node.array = array;
-  } else {
-    this.expectContextual("from");
-    node.array = this.parseMaybeAssign(true);
-  }
+  this.next();
+  this.check(tt.name);
+  const firstIterationVar = this.state.value;
+  this.next();
 
-  this.expectParenFreeBlockStart();
-  node.body = this.parseStatement(false);
-  return this.finishNode(node, "ForFromArrayStatement");
-};
+  let secondIterationType = null, secondIterationVar = null;
+  if (this.match(tt.comma)) {
+    this.next();
+    this.check(tt.name);
+    secondIterationType = this.state.value;
+    if (!(iterableType === getIterableForKeyword(secondIterationType))) {
+      this.raise(this.state.lastTokStart, "Mismatched iteration keywords in for-in loop.");
```

Comment:
might be nice to be more specific, this is probably an area that will trip noobies up...

maybe something like...
```jsx
this.raise(this.state.lastTokStart, `Mismatched iteration keywords: "${firstIterationVar}" is used with ${iterableType}s, while ${secondIterationType} is used with ${getIterableForKeyword(secondIterationType)}s.`);
```